src/sink/buffer.rs

```diff
@@ -0,0 +1,65 @@
+use std::collections::VecDeque;
+
+use {Poll, Async};
+use sink::{Sink, StartSend, AsyncSink};
+
+/// Sink for the `Sink::buffer` combinator, which buffers up to some fixed
+/// number of values when the underlying sink is unable to accept them.
+#[must_use = "sinks do nothing unless polled"]
+pub struct Buffer<S: Sink> {
+    sink: S,
+    buf: VecDeque<S::SinkItem>,
+    cap: usize,
+    need_close: bool,
+}
+
+pub fn new<S: Sink>(sink: S, amt: usize) -> Buffer<S> {
+    Buffer {
+        sink: sink,
+        buf: VecDeque::with_capacity(amt),
+        cap: amt,
+        need_close: false,
+    }
+}
+
+impl<S: Sink> Sink for Buffer<S> {
+    type SinkItem = S::SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        try!(self.poll_complete());
```

Comment:
This seems like it may defeat the purpose, right? This means that you're always trying to write out items, even if you're just starting to push. If the final sink at the bottom is always ready for writing then you'll do syscalls-per-item instead of batching them up into one, right?

Perhaps this could happen only in the case that `self.buf.len() > cap`?